src/http/ws_parser.h

```diff
@@ -0,0 +1,247 @@
+          auto status = serialized::read<uint16_t>(buf, s);
+          auto commit = serialized::read<size_t>(buf, s);
+          auto term = serialized::read<size_t>(buf, s);
+          auto global_commit = serialized::read<size_t>(buf, s);
+
+          std::vector<uint8_t> body(buf, buf + s);
+
+          proc.handle_response(
+            (http_status)status,
+            {{http::headers::CCF_COMMIT, fmt::format("{}", commit)},
+             {http::headers::CCF_TERM, fmt::format("{}", term)},
+             {http::headers::CCF_GLOBAL_COMMIT,
+              fmt::format("{}", global_commit)}},
+            std::move(body));
+          state = INIT;
+          return INITIAL_READ;
+        }
+        default:
+        {
+          throw std::logic_error("Unknown state");
+        }
+      }
+    }
+  };
+
+  class RequestParser : public Parser
+  {
+  private:
+    http::RequestProcessor& proc;
+    uint64_t size = 0;
+    ParserState state = INIT;
+
+  public:
+    RequestParser(http::RequestProcessor& proc_) : proc(proc_) {}
+
+    size_t consume(std::vector<uint8_t>& data) override
+    {
+      switch (state)
+      {
+        case INIT:
+        {
+          assert(data.size() == INITIAL_READ);
+
+          bool fin = data[0] & 0x80;
+          if (!fin)
+          {
+            LOG_FAIL_FMT("Fragment messages aren't supported.");
+            return 0;
+          }
+          else
+          {
+            if (data[0] == 0x82)
+            {
+              if (data[1] & 0x80)
+              {
+                LOG_FAIL_FMT("Masked messages aren't supported.");
+                return 0;
+              }
+              size = data[1] & 0x7f;
+              switch (size)
+              {
+                case 0x7f:
+                {
+                  state = READ_LLEN;
+                  return 8;
+                }
+                case 0x7e:
+                {
+                  state = READ_SLEN;
+                  return 2;
+                }
+                default:
+                {
+                  state = READ_BODY;
+                  return size;
+                }
+              }
+            }
+            else
+            {
+              LOG_FAIL_FMT("Only binary messages are supported.");
+              return 0;
+            }
+          }
+        }
+        case READ_SLEN:
+        {
+          assert(data.size() == 2);
+
+          size = be16toh(*(uint16_t*)data.data());
+          state = READ_BODY;
+          return size;
+        }
+        case READ_LLEN:
+        {
+          assert(data.size() == 8);
+
+          size = be64toh(*(uint64_t*)data.data());
+          state = READ_BODY;
+          return size;
+        }
+        case READ_BODY:
+        {
+          assert(data.size() == size);
+
+          const uint8_t* buf = data.data();
+          size_t s = data.size();
+          auto path = serialized::read_lpsv(buf, s);
+          std::vector<uint8_t> body(buf, buf + s);
+
+          proc.handle_request(
```

Comment:
I guess this means JSON-only payloads with hardcoded `HTTP_POST` only for now?

Reply:
Sort of, in practice the payload can be anything if the handler ignores the content-type, this is just convenient to reuse the logging test right now. Once websocket has it own handler signature, both the verb and the content type will go.